### ISSUE TITLE
FaceElemSolver run_algorithm behavior

### ISSUE THREAD

**Issue by sayerhs** [Exawind/nalu-wind#286]
I am curious about the logic for creation of `ScratchMeInfo` in`AssembleFaceElemSolverAlgorithm::run_algorithm` on GPUs here

https://github.com/Exawind/nalu-wind/blob/da5aa392e078e0705de5cede227b96fdd537b064/include/AssembleFaceElemSolverAlgorithm.h#L48-L59

This snippet of code is being executed in host space and the `MasterElement*` pointers are created on device for _converted_ BC kernels, for example, 

https://github.com/Exawind/nalu-wind/blob/da5aa392e078e0705de5cede227b96fdd537b064/src/edge_kernels/MomentumSymmetryEdgeKernel.C#L39-L42

So my question is, how is line 56 working in the first snippet? The code should crash when executed on host? Yet, CDash indicates that the unit tests work just fine.

**Comment by alanw0** [comment]
I agree it looks like meFC should be a device pointer, so line 56 should seg-fault when attempting to call the 'num_integration_points()' method on host. I'll try to dig into it and see what's actually happening.

**Comment by overfelt** [comment]
In my GPU build the code runs because the Kokkos::DefaultExecutionSpace::memory_space comes up Kokkos::CudaUVMSpace so the GPU allocation of the master elements works fine on the CPU.  Where is this set?  I thought something was set to avoid UVM allocations.

**Comment by alanw0** [comment]
That makes sense. I just added some print statements and saw that the pointer value of the master-element created by create_device_expression was the same pointer that was later used in AssembleFaceElemSolverAlgorithm. A 'MemSpace' value is set in KokkosInterface.h. I don't know about DefaultExecutionSpace::memory_space… That maybe influenced by a cmake configuration parameter for the trilinos build?

**Comment by sayerhs** [comment]
Argh! I should've guessed about the UVM. It is basically required by trilinos Tpetra. So there's no way to disable that I think, unless we can ask Mark if things have changed.

**Comment by alanw0** [comment]
I'll ask Mark about that. He recently told me that tpetra still needs UVM, but I don't know if Kokkos::DefaultExecutionSpace::memory_space has to be UVM because of that.

**Comment by overfelt** [comment]
So if tpetra depends on Kokkos::DefaultExecutionSpace::memory_space being Kokkos::CudaUVMSpace then we would need to define Nalu::DefaultExecutionSpace that would be Kokkos::CudaSpace on the GPU?

**Comment by sayerhs** [comment]
@alanw0 see here https://github.com/kokkos/kokkos/blob/2983b80d9aeafabb81f2c8c1c5a49b40cc0856cb/core/src/Kokkos_Cuda.hpp#L114

**Comment by alanw0** [comment]
I just talked to Mark. He said if we change Kokkos::DefaultExecutionSpace::memory_space we will probably experience pain. But our create_device_expression function is calling kokkos_malloc_on_device which is defined in our KokkosInterface.h header, and it calls Kokkos::kokkos_malloc which takes a template parameter that is defaulting right now. If we pass CudaSpace to that, things should work well. That is Mark's recommendation.

**Comment by sayerhs** [comment]
@alanw0 that sounds like the right thing to do. I have a feeling a lot more segfaults are going to show up once we do that 😉

**Comment by overfelt** [comment]
Yep, kills the NGPKernelTest unit test right off the bat.   I'll look into it....

**Comment by alanw0** [comment]
Yes clearly assembleFaceElemSolverAlgorithm will need some work if we make this change, because line 56 will be segfaulting like Shreyas thought it should be.

**Comment by overfelt** [comment]
The problem seems to be ElemDataRequests and ElemDataRequestsGPU.  ElemDataRequests wants host side element pointers to get information on integration points and such.  ElemDataRequestsGPU is created from ElemDataRequests and copies those pointers but ElemDataRequestsGPU wants to use them on the GPU.  

I don't see enough information contained in the MasterElement base class or the virtual functions defined on it to allow ElemDataRequestsGPU to create GPU side derived class pointer from a host side MasterElement base class pointer. 

Alan, do I understand these two classes correctly?  Or have I missed something....

If this is the case, one solution would be to add a virtual function to MasterElement to return a stk::topology::topology_t that could eventually be used to allocate the correct master element.  Something more direct would be a clone() function.  But maybe there is an easier approach.

**Comment by sayerhs** [comment]
@overfelt You have understood the purposes of `ElemDataRequests` and `ElemDataRequestsGPU` correctly, the latter is supposed to be a _deviceable_ version of the former. Ideally we _register_ the information into `ElemDataRequests` so that we don't have to peek into the `meSCS` instances to get the necessary information. 

This seems to be only necessary for `AssembleFaceElem` where I don't see the real need for `ScratchMeInfo`. We have both `faceDataNeeded` and `elemDataNeeded`, there should be no good reason for us to register a face element in `elemDataNeeded`, @alanw0 do you remember the use case?

I do have [a workaround](https://github.com/sayerhs/nalu-wind/blob/f7612af832f301992d2a6585d5ed01d184257bbc/include/ngp_utils/NgpLoopUtils.h#L94-L128) for this in #287. However, I feel trying to answer _do we really need ScratchMeInfo?_ before workarounds would be the best path forward.

**Comment by alanw0** [comment]
@sayerhs @overfelt I don't specifically remember the use-case, but I'll look into whether we can get rid of ScratchMeInfo.

**Comment by alanw0** [comment]
here's what's going on: before we launch the parallel_for, we pass scratchMeInfo to calculate_shared_mem_bytes_per_thread(..). That function needs several sizes, such as nodes-per-element, nodes-per-face, and num-integration-points for whichever master-elements are being used. We have nodes-per-element and nodes-per-face already, on the host. So we just need a way to get num-integration-points from the device master-elements object. In Shreyas' ngp pull-request he has a 'nodes_per_element' function which gets that from a device master-element. If we have a function like that for num-integration-points, we will be able to get rid of ScratchMeInfo. Or, we could just use that function to get the values to put into ScratchMeInfo.

**Comment by sayerhs** [comment]
@alanw0 It would be good to get rid of the `ScratchMeInfo` object if possible. Will something [like this](https://github.com/Exawind/nalu-wind/pull/287/files#diff-b322bf8d90f3c2a2474980ac868f3920R79) in the loop abstractions pull request (#287) be sufficient?

**Comment by sayerhs** [comment]
I'll add the `num_integration_points()` utility function like the `nodes_per_element()`

**Comment by alanw0** [comment]
@sayerhs Do you already have something equivalent to the 'get_num_scalars_pre_req_data' function at line 332 of ScratchViews.C? That's where it figures out the sizes for the master-element-views such as gradop, etc.

**Comment by sayerhs** [comment]
I am using functions defined in `ScratchViews.C` -- [see here](https://github.com/Exawind/nalu-wind/pull/287/files#diff-b322bf8d90f3c2a2474980ac868f3920R55)

**Comment by alanw0** [comment]
@sayerhs I don't see where it calculates the sizes for master-element views such as gradop or areav etc.

**Comment by sayerhs** [comment]
@alanw0 I am using this function defined in `ScratchViews.C` that calls `get_num_scalars_pre_req_data` internally 

https://github.com/Exawind/nalu-wind/blob/718131130dc75e4da66fd64999b2ad3c07586237/include/ScratchViews.h#L919-L922

```C++
inline int ngp_calc_thread_shmem_size(int ndim, const DataReqType& dataReq)
	{
	  int preReqSize = get_num_bytes_pre_req_data<T>(dataReq, ndim);
	  int mdvSize = MultiDimViews<T>::bytes_needed(
	    dataReq.get_total_num_fields(), count_needed_field_views(dataReq));
	
	#ifndef KOKKOS_ENABLE_CUDA
	  // On host account for extra data to store the SIMD and non-SIMD versions of
	  // the ScratchViews
	  return (preReqSize + mdvSize) * 2 * simdLen;
	#else
	  return (preReqSize + mdvSize);
	#endif
	}
```

**Comment by alanw0** [comment]
@sayerhs Ah ha, ok. I hadn't noticed (or had forgotten about) that overload of get_num_scalars_pre_req_data that doesn't take a ScratchMeInfo. That's essentially a duplicated function, it looks like we could delete the one that takes ScratchMeInfo, and then get rid of ScratchMeInfo.

**Comment by sayerhs** [comment]
@alanw0 Looking further into the code, it appears to me that `ScratchMeInfo` exists solely to pass `numFaceIp` to the code that processes `elemDataNeeded` (where the user might have requested `SCS_FACE_GRAD_OP`, so instead of `meInfo` we can just pass this extra parameter to the functions? 

Also there is a bug in the current implementation of how we compute the scratch memory size for `elemDataNeeded` from within `AssembleFaceElem`. Currently we extract the `numScsIp` and `numScvIp` from master elements registered to `faceDataNeeded` 

https://github.com/Exawind/nalu-wind/blob/89f5fb190b4d87695e8398b0c69dfa5d94d3dd25/include/AssembleFaceElemSolverAlgorithm.h#L48-L58

and, if I understand the design intent correctly, `meSCS` and `meSCV` are never registered to `faceDataNeeded`, but `elemDataNeeded` instead. So in `ScratchMeInfo`, `numScsIp = numScvIp = 0`. Now this `meInfo` object is passed to `get_num_scalars_pre_req_data` along with `elemDataNeeded` object 

https://github.com/Exawind/nalu-wind/blob/89f5fb190b4d87695e8398b0c69dfa5d94d3dd25/include/ScratchViews.h#L956-L971

where, instead of consulting `meSCS` and `meSCV` registered to `elemDataNeeded`, it pulls `numScsIp` and `numScvIp` from `meInfo` object:

https://github.com/Exawind/nalu-wind/blob/89f5fb190b4d87695e8398b0c69dfa5d94d3dd25/src/ScratchViews.C#L332-L338

and uses that to compute sizes: 

https://github.com/Exawind/nalu-wind/blob/89f5fb190b4d87695e8398b0c69dfa5d94d3dd25/src/ScratchViews.C#L388-L401 

I think so far we have been lucky that there wasn't a use-case where we would request anything other than `SCS_FACE_GRAD_OP` or `SCS_SHIFTED_FACE_GRAD_OP` from `meSCS` within a `FaceElem` algorithm.

**Comment by alanw0** [comment]
@sayerhs I'm glad you're unraveling this stuff, I think your changes are a great improvement.

**Comment by sayerhs** [comment]
Fixed with #293 and #294
